core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java

```diff
@@ -137,34 +145,44 @@ void readUnread(@Nullable Snapshot.Key key, Get get) throws CrudException {
   @VisibleForTesting
   Optional<TransactionResult> read(@Nullable Snapshot.Key key, Get get) throws CrudException {
     Optional<TransactionResult> result = getFromStorage(get);
-    if (!result.isPresent() || result.get().isCommitted()) {
-      if (result.isPresent() || get.getConjunctions().isEmpty()) {
-        // Keep the read set latest to create before image by using the latest record (result)
-        // because another conflicting transaction might have updated the record after this
-        // transaction read it first. However, we update it only if a get operation has no
-        // conjunction or the result exists. This is because we don’t know whether the record
-        // actually exists or not due to the conjunction.
-        if (key != null) {
-          putIntoReadSetInSnapshot(key, result);
-        } else {
-          // Only for a Get with index, the argument `key` is null
+    if (result.isPresent() && !result.get().isCommitted()) {
+      throw new UncommittedRecordException(
+          get,
+          result.get(),
+          CoreError.CONSENSUS_COMMIT_READ_UNCOMMITTED_RECORD.buildMessage(),
+          snapshot.getId());
+    }
 
-          if (result.isPresent()) {
-            // Only when we can get the record with the Get with index, we can put it into the read
-            // set
-            key = new Snapshot.Key(get, result.get());
-            putIntoReadSetInSnapshot(key, result);
-          }
+    if (!get.getConjunctions().isEmpty()) {
+      // Check if the result matches the conjunctions
```

Comment:
It's likely this check exists because we get records based on not only the current image but also before image using OR conditions. We would like to filter the case where the second record is committed in the example of the Javadoc comment. If my understanding is correct, it might be nice to have a comment like this for future readers to easily grasp the reasoning.

```suggestion
      // Because we also get records whose before images match the conjunctions, we need to check if
      // the current status of the records actually match the conjunctions.
```

Reply:
Fixed in 4ec25d631eab49f358a7068b7af68ac854602648. Thanks!